OrbitCore/LinuxEventTracer.cpp

```diff
@@ -99,150 +109,229 @@ void LinuxEventTracer::Run(
+              sample.SetFunction(
+                  uprobe_fds_to_function.at(ring_buffer.FileDescriptor()));
+              uprobe_event_processor.Push(
+                  std::make_unique<LinuxUprobeEventWithStack>(
+                      std::move(sample)));
+
+            } else if (is_uretprobes) {
+              auto sample =
+                  ring_buffer.ConsumeRecord<LinuxUretprobeEventWithStack>(
+                      header);
+
+              std::vector<unwindstack::FrameData> frames = unwinder.Unwind(
+                  sample.Registers(), sample.StackDump(), sample.StackSize());
+              if (!frames.empty()) {
+                HandleCallstack(sample.TID(), sample.Timestamp(), frames);
+              }
+
+              sample.SetFunction(
+                  uretprobe_fds_to_function.at(ring_buffer.FileDescriptor()));
+              uprobe_event_processor.Push(
+                  std::make_unique<LinuxUretprobeEventWithStack>(
+                      std::move(sample)));
 
-            std::vector<unwindstack::FrameData> frames = unwinder.Unwind(
-                sample.Registers(), sample.StackDump(), sample.StackSize());
-            if (!frames.empty()) {
-              HandleCallstack(sample.TID(), sample.Timestamp(), frames);
+            } else {
+              auto sample =
+                  ring_buffer.ConsumeRecord<LinuxStackSampleEvent>(header);
+
+              std::vector<unwindstack::FrameData> frames = unwinder.Unwind(
+                  sample.Registers(), sample.StackDump(), sample.StackSize());
+              if (!frames.empty()) {
+                HandleCallstack(sample.TID(), sample.Timestamp(), frames);
+              }
             }
           } break;
 
           case PERF_RECORD_LOST: {
             auto lost = ring_buffer.ConsumeRecord<LinuxPerfLostEvent>(header);
-            PRINT("Lost %u Events\n", lost.Lost());
+            PRINT("Lost %u events\n", lost.Lost());
           } break;
 
-          default:
-            PRINT("Unexpected Perf Sample Type: %u\n", header.type);
+          default: {
+            PRINT("Unexpected perf_event_header::type: %u\n", header.type);
             ring_buffer.SkipRecord(header);
-            break;
+          } break;
         }
       }
     }
+
+    if (new_thread >= 0) {
+      int32_t fd =
+          LinuxPerfUtils::stack_sample_event_open(new_thread, period_ns_);
+      fds.push_back(fd);
+      ring_buffers.emplace_back(fd);
+      LinuxPerfUtils::start_capturing(fd);
+    }
+
+    uprobe_event_processor.ProcessTillOffset();
   }
 
+  uprobe_event_processor.ProcessAll();
+
   // stop capturing
   for (int32_t fd : fds) {
     LinuxPerfUtils::stop_capturing(fd);
   }
+
   ring_buffers.clear();
   for (int32_t fd : fds) {
     close(fd);
   }
 }
 
-//-----------------------------------------------------------------------------
-void LinuxEventTracer::HandleCallstack(
-    ThreadID tid, uint64_t timestamp,
+bool LinuxEventTracerThread::ComputeFrequency() {
```

Comment:
The name is misleading. It is not computing the frequency, the frequency is given (per frequency_ member), it is computing the sampling period (in ns). 

I would also consider making this a purely functional signature (no member access), like:

bool ComputePeriodInNs(double frequency, uint64_t* period_ns);

Reply:
Oh yeah, the name is wrong.
As for making it purely functional, I'm not sure what the benefit would be at the moment as the result would always be assigned to the same member of this class.